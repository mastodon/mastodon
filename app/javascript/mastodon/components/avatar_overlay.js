import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { autoPlayGif } from '../initial_state';
import Avatar from './avatar';

export default class AvatarOverlay extends React.PureComponent {

  static propTypes = {
    account: ImmutablePropTypes.map.isRequired,
    friend: ImmutablePropTypes.map.isRequired,
    animate: PropTypes.bool,
    size: PropTypes.number,
    baseSize: PropTypes.number,
    overlaySize: PropTypes.number,
  };

  static defaultProps = {
    animate: autoPlayGif,
    size: 46,
    baseSize: 36,
    overlaySize: 24,
  };

  state = {
    hovering: false,
  };

  handleMouseEnter = () => {
    if (this.props.animate) return;
    this.setState({ hovering: true });
  };

  handleMouseLeave = () => {
    if (this.props.animate) return;
    this.setState({ hovering: false });
  };

  render() {
    const { account, friend, animate, size, baseSize, overlaySize } = this.props;
    const { hovering } = this.state;

    return (
      <div className='account__avatar-overlay' style={{ width: size, height: size }}>
        <div className='account__avatar-overlay-base'><Avatar animate={hovering || animate} account={account} size={baseSize} /></div>
        <div className='account__avatar-overlay-overlay'><Avatar animate={hovering || animate} account={friend} size={overlaySize} /></div>
      </div>
    );
  }

}
