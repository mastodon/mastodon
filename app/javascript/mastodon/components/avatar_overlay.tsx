import type { Account } from 'mastodon/models/account';

import { useHovering } from '../../hooks/useHovering';
import { autoPlayGif } from '../initial_state';

interface Props {
  account: Account | undefined; // FIXME: remove `undefined` once we know for sure its always there
  friend: Account | undefined; // FIXME: remove `undefined` once we know for sure its always there
  size?: number;
  baseSize?: number;
  overlaySize?: number;
}

export const AvatarOverlay: React.FC<Props> = ({
  account,
  friend,
  size = 46,
  baseSize = 36,
  overlaySize = 24,
}) => {
  const { hovering, handleMouseEnter, handleMouseLeave } =
    useHovering(autoPlayGif);
  const accountSrc = hovering
    ? account?.get('avatar')
    : account?.get('avatar_static');
  const friendSrc = hovering
    ? friend?.get('avatar')
    : friend?.get('avatar_static');

  return (
    <div
      className='account__avatar-overlay'
      style={{ width: size, height: size }}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      <div className='account__avatar-overlay-base'>
        <div
          className='account__avatar'
          style={{ width: `${baseSize}px`, height: `${baseSize}px` }}
        >
          {accountSrc && <img src={accountSrc} alt={account?.get('acct')} />}
        </div>
      </div>
      <div className='account__avatar-overlay-overlay'>
        <div
          className='account__avatar'
          style={{ width: `${overlaySize}px`, height: `${overlaySize}px` }}
        >
          {friendSrc && <img src={friendSrc} alt={friend?.get('acct')} />}
        </div>
      </div>
    </div>
  );
};
