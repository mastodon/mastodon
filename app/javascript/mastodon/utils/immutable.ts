import { isImmutable } from 'immutable';

/**
 * Converts an immutable Map or Set to JS if it isn't already a plain JS object.
 * This is useful in code that needs to be able to deal with either type.
 */

export function immutableToJSIfNeeded<T>(object: T): T {
  return isImmutable(object) ? (object.toJS() as T) : object;
}
