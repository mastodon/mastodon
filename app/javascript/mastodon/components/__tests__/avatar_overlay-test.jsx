import { fromJS } from 'immutable';

import renderer from 'react-test-renderer';

import { AvatarOverlay } from '../avatar_overlay';

describe('<AvatarOverlay', () => {
  const account = fromJS({
    username: 'alice',
    acct: 'alice',
    display_name: 'Alice',
    avatar: '/animated/alice.gif',
    avatar_static: '/static/alice.jpg',
  });

  const friend = fromJS({
    username: 'eve',
    acct: 'eve@blackhat.lair',
    display_name: 'Evelyn',
    avatar: '/animated/eve.gif',
    avatar_static: '/static/eve.jpg',
  });

  it('renders a overlay avatar', () => {
    const component = renderer.create(<AvatarOverlay account={account} friend={friend} />);
    const tree      = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
