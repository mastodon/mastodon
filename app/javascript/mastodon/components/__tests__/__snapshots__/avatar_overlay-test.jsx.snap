// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarOverlay renders a overlay avatar 1`] = `
<div
  className="account__avatar-overlay"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    {
      "height": 46,
      "width": 46,
    }
  }
>
  <div
    className="account__avatar-overlay-base"
  >
    <div
      className="account__avatar"
      style={
        {
          "height": "36px",
          "width": "36px",
        }
      }
    >
      <img
        alt="alice"
        src="/static/alice.jpg"
      />
    </div>
  </div>
  <div
    className="account__avatar-overlay-overlay"
  >
    <div
      className="account__avatar"
      style={
        {
          "height": "24px",
          "width": "24px",
        }
      }
    >
      <img
        alt="eve@blackhat.lair"
        src="/static/eve.jpg"
      />
    </div>
  </div>
</div>
`;
